lib/functions/RegularFunctions.ts

```diff
@@ -165,11 +165,10 @@ const equality = {
   arity: 2,
   overloads: declare(C.RegularOperator.EQUAL)
     .numberTest(() => (left, right) => left === right)
-    .stringTest(() => (left, right) => left.localeCompare(right) === 0)
     .set(
-      [ TypeURL.RDF_LANG_STRING, TypeURL.RDF_LANG_STRING ],
-      () => ([ left, right ]: E.LangStringLiteral[]) => bool(left.str() === right.str() &&
-        left.language === right.language),
+      [ TypeAlias.SPARQL_STRINGLY, TypeAlias.SPARQL_STRINGLY ],
+      () => ([ left, right ]: (E.LangStringLiteral | E.StringLiteral)[]) => bool(left.str() === right.str() &&
+        left.language === right.language && left.dataType === right.dataType),
```

Comment:
Note: I added the datatype check here. But I'd need to check what the spec says about this.

Reply:
Makes sense to me!

Reply:
Yes, the only problem is, do we want this for our type discovery? What happens if we evaluate `"test"^^ex:myStr == "test"^^xsd:string` and we discovered that `ex:myStr` is a subtype of `xsd:string`? I think, in this case, the terms are equal? 

I think the spec says the same, but I am not 100% sure if I interoperate it correctly, could you verify?
In [Operand Data Types](https://www.w3.org/TR/2013/REC-sparql11-query-20130321/#operandDataTypes) the last sentence:
> SPARQL language extensions may treat additional types as being derived from XML schema datatypes.

I think this means: If you get `ex:myStr`, a subtype of `xsd:string`, you can use the following functions as if it was an `xsd:string`.

Reply:
No, AFAIK, literals with unequal datatype are never equal.

The part of the spec you refer to talks about acceptable function args, for which subtypes may be considered.

Reply:
I don't think this is true, although it makes sense... When checking for equality using [term equal](https://www.w3.org/TR/2013/REC-sparql11-query-20130321/#OperatorMapping), it is the case that both types need to be the same.
However, the [operator mapping part](https://www.w3.org/TR/2013/REC-sparql11-query-20130321/#OperatorMapping) just says subtype substitution needs to be handled. If subtype substitution needs to be handled, I think my example above should result in true.

Reply:
Hmm, good catch, I think you're right.

So this statement is correct then:

> What happens if we evaluate "test"^^ex:myStr == "test"^^xsd:string and we discovered that ex:myStr is a subtype of xsd:string? I think, in this case, the terms are equal?

Do you think this will break with this PR?

Reply:
As it is now, It will break, but I will make a commit later so everything is correct.